### PR TITLE
[🍒][cxx-interop] Only mark projections of self-contained types as unsafe.

### PR DESCRIPTION
**Pull request by zoecarver**
**Description:** Projections of trivial types and view types aren't unsafe. This matches what was described in the vision document. A view type like StringRef is already referencing memory it does not own, so if it projects its data again (via the data method for example), there's no additional unsafety. The really unsafe thing is when a self contained (owned) type projects its owned memory because there are subtle difference in C++ and Swift lifetime models (along with different safety expectations). I have discussed with the interop code owners and we all agree that this is the correct change to make. This has been corroborated by feedback from adopters. Finally, this is the behavior that was agreed upon by the workgroup and the broader community and is what is outlined in the forward vision document. 
**Scope:** C++ interop
**Risk:** Medium. This change will break source. It changes an important part of importing and affects safety. It also has a high risk of breaking parts of the overlay. This change will be hard to walk back if we determine that it is unsafe (for example). It has not been testing with any adopters and this new behavior has not been tested in the wild.
**Testing:** Unit testing (CI)
**Original PR:** #67296
**Radar:** rdar://112220411

### CODE REVIEW

**Comment by zoecarver** [comment]
@swift-ci please test

**Comment by zoecarver** [comment]
@swift-ci Please Build Toolchain macOS Platform

File: SwiftCompilerSources/Sources/Basic/Utils.swift [hunk 1]
```diff
@@ -64,11 +64,34 @@ public struct StringRef : CustomStringConvertible, NoReflectionChildren {
...
+#endif
+  }
+
+  public subscript(index: Int) -> UInt8 {
+#if $NewCxxMethodSafetyHeuristics
+    let buffer = UnsafeBufferPointer<UInt8>(start: _bridged.bytes_begin(),
+                                            count: count)
+#else
+    let buffer = UnsafeBufferPointer<UInt8>(start: _bridged.bytes_begin(),
+                                            count: count)
```
Comment (ravikandhadai) [inline SwiftCompilerSources/Sources/Basic/Utils.swift:82]: @zoecarver both #if and #else branches seems to have the same code? Did we mean to say `_bridged.__bytes_beginUnsafe()` in the #else case?

File: lib/ClangImporter/ClangImporter.cpp [hunk 2]
```diff
@@ -6778,20 +6823,22 @@ bool IsSafeUseOfCxxDecl::evaluate(Evaluator &evaluator,
...
-        }
 
         // Mark this as safe to help our diganostics down the road.
         if (!cxxRecordReturnType->getDefinition()) {
           return true;
         }
 
-        if (!hasCustomCopyOrMoveConstructor(cxxRecordReturnType) &&
+        // A projection of a view type (such as a string_view) from a self
+        // contained parent is a proejction (unsafe).
```
Comment (ravikandhadai) [inline lib/ClangImporter/ClangImporter.cpp:6837]: Minor typo: "proejction", we can combine the fix with later fixes.

**Comment by zoecarver** [comment]
@swift-ci please test

**Comment by zoecarver** [comment]
@swift-ci Please Build Toolchain macOS Platform

**Comment by zoecarver** [comment]
@swift-ci Please Build Toolchain macOS Platform

**Comment by zoecarver** [comment]
@swift-ci please test

**Comment by zoecarver** [comment]
@swift-ci please test macOS

**Comment by zoecarver** [comment]
@swift-ci please test windows

**Review by ravikandhadai — APPROVED** [review]
Thanks Zoe! LGTM.

**Pull request merged**